### PR TITLE
[full-ci] Add global loading indicator for long running tasks

### PR DESCRIPTION
**Pull request by JammingBen**
## Description
Adds a global loading indicator for long running actions at the top of the page.

## How to use it (for devs)

Simply wrap your Promise in the `addTask` method provided by the `loadingService`:

```
loadingService.addTask(() => {
  return new Promise((resolve, reject) => {
    resolve(true)
  })
})
```

The method also provides a callback `setProgress` which can be used to update the progress of a task.

```
loadingService.addTask((setProgress) => {
  const promises = []
  const actions = [1, 2, 3, 4, 5]

  for (const [i, action] of actions.entries()) {
    promises.push(
      new Promise((resolve, reject) => {
        setProgress({ total: actions.length, current: i + 1 })
        resolve(true)
      })
    )
  }

  return Promise.all(promises)
}, { indeterminate: false })
```

The following options can be provided to `addTask`:

* `debounceTime` - Time until the loading state for a task is set active. Defaults to `200` ms
* `indeterminate` - Determines if the task is in an indeterminate state. Defaults to `true`.

Note that the loading bar won't show a progress when at least one task is in an indeterminate state.

Thx to @dschmidt for helping me out with the task adding & handling!

## Video

https://user-images.githubusercontent.com/50302941/225225397-f99075c4-859b-49da-8b13-66d4fc422188.mov

## Related Issue

- Fixes https://github.com/owncloud/web/issues/6183
- Fixes https://github.com/owncloud/web/issues/2134

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

### CODE REVIEW

**Comment by ownclouders** [comment]
Results for <strong>acceptance oC10</strong> https://drone.owncloud.com/owncloud/web/33656/32/1
The following scenarios passed on retry:
- webUITrashbinDelete/trashbinDelete.feature:36
<details><summary>:boom: The acceptance tests failed on retry. Please find the screenshots inside ...</summary>

<p>


### webUITrashbinDelete-trashbinDelete_feature-L110.png
  

### webUITrashbinDelete-trashbinDelete_feature-L123.png
  

### webUITrashbinDelete-trashbinDelete_feature-L85.png
  


</p></details>
:boom: The oC10Trashbin tests pipeline failed. The build has been cancelled.

File: packages/web-pkg/src/services/loadingService.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    if (!task.state) {
+      task.state = { total: 0, current: 0 }
+    }
+    task.state.total = total
+    task.state.current = current
+    eventBus.publish(LoadingEventTopics.setProgress)
+  }
+}
+
+export const loadingService = new LoadingService()
```
Comment (dschmidt) [inline packages/web-pkg/src/services/loadingService.ts:100]: I would prefer not to export a "default instance", but to make it available via `app.config.globalProperties.$loadingService` and to access that through a `useLoadingService` composable

Yes, I know we create these default instances for all or at least most of the other services, but I still think it's pretty ugly and we should stop 😅

File: packages/web-pkg/tests/unit/services/loadingService.spec.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    expect(service.isLoading).toBeFalsy()
+  })
+  it('adds a task and sets it active after the debounce', () => {
+    const service = new LoadingService()
+    const action = new Promise((resolve) => {
+      resolve(true)
+    })
+
+    service.addTask(() => action)
+    jest.runAllTimers()
```
Comment (dschmidt) [inline packages/web-pkg/tests/unit/services/loadingService.spec.ts:23]: Maybe add another check *before* the timers are run to check `isLoading` is falsy?

File: packages/web-pkg/src/services/loadingService.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    return Math.round((progress / tasks.length) * 100)
+  }
+
+  public addTask(
+    callback: ({ setProgress }: { setProgress: (args: LoadingTaskState) => void }) => Promise<any>,
+    {
+      debounceTime = DEFAULT_DEBOUNCE_TIME,
+      indeterminate = true
+    }: { debounceTime?: number; indeterminate?: boolean } = {}
+  ): Promise<any> {
```
Comment (dschmidt) [inline packages/web-pkg/src/services/loadingService.ts:60]: I think, we can improve the types a little:

```typescript
// not a great name, better ideas appreciated
// maybe `LoadingTaskProgress` and then a `.set` method?
export interface LoadingTaskCallbackArguments {
  setProgress: (args: LoadingTaskState) => void
}
// …
  public addTask<T>(
    callback: ({ setProgress }: LoadingTaskCallbackArguments) => Promise<T>,
    {
      debounceTime = DEFAULT_DEBOUNCE_TIME,
      indeterminate = true
    }: { debounceTime?: number; indeterminate?: boolean } = {}
  ): Promise<T> {
  // …
  }
```

That way 
- we can more easily pass on the `LoadingTaskCallbackArguments` to helper classes/functions, which I prefer over just passing around the `setProgress` function, as it makes it easier to add more features later (`.advanceOne()`, `.advance(2)` or something like that e.g.  :)) 
- we (more importantly) don't lose the return type of the callback function

**Review by dschmidt — COMMENTED** [review]
I really like where this is going - thanks for taking the time to try the "`addTask` approach" I suggested.

File: packages/web-pkg/tests/unit/services/loadingService.spec.ts (hunk 2 above)
Comment (JammingBen) [reply packages/web-pkg/tests/unit/services/loadingService.spec.ts:23]: I already have a separate test that covers exactly this scenario. I like to have that more explicitly :)

**Comment by tbsbdr** [comment]
Visual improvement proposal
- @JammingBen et al: movement catches our attention by itself; thus I'd propose to make the bar more decent; do you think this would be an improvement?


https://user-images.githubusercontent.com/26610733/225310499-fc3242e1-ecf8-4f74-8c87-cc5f8291786d.mp4

**Review by dschmidt — APPROVED** [review]

File: packages/web-app-files/src/store/actions.ts [hunk 4]
```diff
@@ -205,6 +204,9 @@ export default {
             { root: true }
           )
         })
+        .finally(() => {
+          setProgress({ total: files.length, current: removedFiles.length })
```
Comment (dschmidt) [inline packages/web-app-files/src/store/actions.ts:211]: Hm - looking at this again: a file is pushed only to the removedFiles after success, right? So just moving this to finally should have no effect I guess.

Comment (JammingBen) [reply packages/web-app-files/src/store/actions.ts:211]: Or simply do it like in the other places via `for (const [i, file] of files.entries())` and use `i`?

File: packages/web-app-files/src/composables/actions/files/useFileActionsRestore.ts [hunk 5]
```diff
@@ -256,7 +262,12 @@ export const useFileActionsRestore = ({ store }: { store?: Store<any> } = {}) =>
       resource.path = urlJoin(parentPath, resolvedName)
       resolvedResources.push(resource)
     }
-    return restoreResources(space, resolvedResources, missingFolderPaths)
+    return loadingService.addTask(
+      ({ setProgress }) => {
+        return restoreResources(space, resolvedResources, missingFolderPaths, { setProgress })
+      },
```
Comment (dschmidt) [inline packages/web-app-files/src/composables/actions/files/useFileActionsRestore.ts:268]: No need to destructure and build a new object, but not so important

**Review by dschmidt — APPROVED** [review]

**Pull request merged**